modules/vstudio/vs2010_vcxproj.lua

```diff
@@ -3666,6 +3970,451 @@
+		if cfg.strictaliasing ~= nil then
+			m.element("StrictAliasing", nil, iif(cfg.strictaliasing == "Off", "false", "true"))
+		end
+	end
+
+	function m.androidAPILevel(cfg)
+		if cfg.androidapilevel ~= nil then
+			m.element("AndroidAPILevel", nil, "android-" .. cfg.androidapilevel)
+		end
+	end
+
+	function m.androidFpu(cfg)
+		if cfg.fpu ~= nil then
+			-- TODO REVIEW THIS
+			_p(3,'<SoftFloat>true</SoftFloat>', iif(cfg.fpu == "Software", "true", "false"))
+		end
+	end
+
+	function m.androidShortEnums(cfg)
+		if cfg.flags.UseShortEnums ~= nil then
+			m.element(UseShortEnums, nil, "true")
+		end
+	end
+
+	function m.androidDataLevelLinking(cfg)
+		-- Enables linker optimizations to remove unused data by emitting each data item in a separate section
+		if cfg.flags.DataLevelLinking ~= nil then
+			m.element("DataLevelLinking", nil, "true")
+		end
+	end
+
+	function m.thumbMode(cfg)
+		if cfg.thumbmode ~= nil then
+			local thumbMode =
+			{
+				thumb = "Thumb",
+				arm = "ARM",
+				disabled = "Disabled",
+			}
+			m.element("ThumbMode", nil, thumbMode[cfg.thumbmode])
+		end
+	end
+
+	function m.androidGenerateMapFile(cfg)
+		if cfg.flags.Maps then
+			-- Android specifies a name. Other platforms use the project name
+			-- so we do the same thing here
+			m.element("GenerateMapFile", nil, cfg.project.name..".map")
+		end
+	end
+
+	function m.androidUseMultiToolTask(cfg)
+		-- Android equivalent of 'MultiProcessorCompilation'
+		if cfg.flags.MultiProcessorCompile then
+			m.element("UseMultiToolTask", nil, "true")
+		end
+	end
+
+	-- Remove .lib and list in LibraryDependencies instead of AdditionalDependencies
+	function m.androidAdditionalDependencies(cfg, explicit)
+
+		local links = {}
+
+		-- If we need sibling projects to be listed explicitly, grab them first
+		if explicit then
+			links = config.getlinks(cfg, "siblings", "fullpath")
+		end
+
+		-- Then the system libraries, which come undecorated
+		local system = config.getlinks(cfg, "system", "name")
+		for i = 1, #system do
+			local link = system[i]
+			table.insert(links, link)
+		end
+
+		-- TODO: When to use LibraryDependencies vs AdditionalDependencies
+
+		if #links > 0 then
+			links = path.translate(table.concat(links, ";"))
+			m.element("LibraryDependencies", nil, "%%(LibraryDependencies);%s", links)
+		end
+
+	end
+
+	function m.androidAntBuildPreVS2015(cfg)
+		if cfg.kind == premake.STATICLIB or cfg.kind == premake.SHAREDLIB then
+			return
+		end
+
+		_p(2,'<AntBuild>')
+		_p(3,'<AntBuildType>%s</AntBuildType>', iif(premake.config.isDebugBuild(cfg), "Debug", "Release"))
+		_p(2,'</AntBuild>')
+	end
+
+	--
+	-- Shared project generation functions
+	--
+
+	function m.androidLinuxIncludePath(cfg)
+		-- Disable usage of ExternalIncludePath, Android project format ignores it
+		local dirs = vstudio.path(cfg, cfg.externalincludedirs)
+		if #dirs > 0 then
+			m.element("IncludePath", nil, "%s;$(IncludePath)", table.concat(dirs, ";"))
+		end
+	end
```

Comment:
Not sure I agree with this function. The way it works isn't wrong, but extending the original function is probably better. Then just add a function like:
```lua
function m.hasExternalIncludePath(cfg)
  if ... then -- The inverse of the current check
    return true
  end

  return false
end
```
This function allows the extensibility without requiring the entire function contents to be duplicated.

Reply:
Hmm that won't have the same behavior. This is an interesting thing though, it seems as though in e.g. VS2019 these tags are both allowed, I can have IncludePath and ExternalIncludePath, but the existing common code won't allow it. Do you know the rationale for this?

```lua
if #dirs > 0 then
	if _ACTION < "vs2019" then
		m.element("IncludePath", nil, "%s;$(IncludePath)", table.concat(dirs, ";"))
	else
		m.element("ExternalIncludePath", nil, "%s;$(ExternalIncludePath)", table.concat(dirs, ";"))
	end
end
```

Could you suggest how to merge both of these functions into something that makes sense to you?

Reply:
On closer study I think I just need to delete the androidLinuxIncludePath and just add the regular one to the Android project so it supports the externalincludedirs api, I'll do that right away

Reply:
The issue here is that Android projects don't support the `ExternalIncludePath` element, so projects with `includedirs` or `externalincludedirs` both fail to build. VS2022 Android projects might support it now, but I'm fairly certain they didn't before - I submitted the override to make Android projects keep using `IncludePath`.

I don't know why for VS2019 and above we only use `ExternalIncludePath` instead of splitting the two lists across the two elements.

> Hmm that won't have the same behavior. 

It should, but I may not have explained my suggestion well enough. Changing the `if _ACTION < "vs2019" then` to `if m.hasExternalIncludePath(cfg) then`, you can make that function return `_ACTION < "vs2019" or cfg.system == p.LINUX or cfg.system == p.ANDROID`. This helper function allows users and modules to easily override the behaviour when/if required.

Reply:
I see what you mean and it could be useful. However in this instance, in the interest of simplicity, I have confirmed that Android VS2019 and 2022 both support the external directory, and VS2017 does not. Therefore the default behavior of the VS module is correct and doesn't need the patched version